src/source-dpdk.c

```diff
@@ -0,0 +1,617 @@
+ * @param real_tv
+ */
+static void DPDKSetTimevalReal(struct timeval *machine_start_tv, struct timeval *real_tv)
+{
+    CyclesAddToTimeval(rte_get_tsc_cycles(), machine_start_tv, real_tv);
+}
+
+/* get number of seconds from the reset of TSC counter (typically from the machine start) */
+static uint64_t DPDKGetSeconds()
+{
+    return CyclesToSeconds(rte_get_tsc_cycles());
+}
+
+/**
+ * Attempts to retrieve NUMA node id on which the caller runs
+ * @return NUMA id on success, -1 otherwise
+ */
+static int GetNumaNode(void)
+{
+    int cpu = 0;
+    int node = -1;
+
+#if defined(__linux__)
+    cpu = sched_getcpu();
+    node = numa_node_of_cpu(cpu);
+#else
+    SCLogWarning(SC_ERR_TM_THREADS_ERROR, "NUMA node retrieval is not supported on this OS.");
+#endif
+
+    return node;
+}
+
+/**
+ * \brief Registration Function for ReceiveDPDK.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleReceiveDPDKRegister(void)
+{
+    tmm_modules[TMM_RECEIVEDPDK].name = "ReceiveDPDK";
+    tmm_modules[TMM_RECEIVEDPDK].ThreadInit = ReceiveDPDKThreadInit;
+    tmm_modules[TMM_RECEIVEDPDK].Func = NULL;
+    tmm_modules[TMM_RECEIVEDPDK].PktAcqLoop = ReceiveDPDKLoop;
+    tmm_modules[TMM_RECEIVEDPDK].PktAcqBreakLoop = NULL;
+    tmm_modules[TMM_RECEIVEDPDK].ThreadExitPrintStats = ReceiveDPDKThreadExitStats;
+    tmm_modules[TMM_RECEIVEDPDK].ThreadDeinit = ReceiveDPDKThreadDeinit;
+    tmm_modules[TMM_RECEIVEDPDK].cap_flags = SC_CAP_NET_RAW;
+    tmm_modules[TMM_RECEIVEDPDK].flags = TM_FLAG_RECEIVE_TM;
+}
+
+/**
+ * \brief Registration Function for DecodeDPDK.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleDecodeDPDKRegister(void)
+{
+    tmm_modules[TMM_DECODEDPDK].name = "DecodeDPDK";
+    tmm_modules[TMM_DECODEDPDK].ThreadInit = DecodeDPDKThreadInit;
+    tmm_modules[TMM_DECODEDPDK].Func = DecodeDPDK;
+    tmm_modules[TMM_DECODEDPDK].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_DECODEDPDK].ThreadDeinit = DecodeDPDKThreadDeinit;
+    tmm_modules[TMM_DECODEDPDK].cap_flags = 0;
+    tmm_modules[TMM_DECODEDPDK].flags = TM_FLAG_DECODE_TM;
+}
+
+static inline void DPDKDumpCounters(DPDKThreadVars *ptv)
+{
+    struct rte_eth_stats eth_stats;
+    int retval = rte_eth_stats_get(ptv->port_id, &eth_stats);
+    if (unlikely(retval != 0)) {
+        SCLogError(SC_ERR_STAT, "Failed to get stats for port id %d: %s", ptv->port_id,
+                strerror(-retval));
+        return;
+    }
+
+    uint64_t th_pkts = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_packets);
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_packets, ptv->pkts - th_pkts);
+    SC_ATOMIC_ADD(ptv->livedev->pkts, ptv->pkts - th_pkts);
+
+    /* Some NICs (e.g. Intel) do not support queue statistics and the drops can be fetched only on
+     * the port level. Therefore setting it to the first worker to have at least continuous update
+     * on the dropped packets. */
+    if (ptv->queue_id == 0) {
+        StatsSetUI64(ptv->tv, ptv->capture_dpdk_rx_errs,
```

Comment:
I think we should get the drop counter back here. `rx_errors` seems to include it, but we need to follow the other methods on this.

I am thinking about capture method specific stats. Perhaps we can have `capture.dpdk.rx_nombuf`, and such.

Reply:
This is probably all drops, but its hard to know for sure
```
Date: 11/14/2021 -- 13:52:45 (uptime: 0d, 04h 11m 41s)
------------------------------------------------------------------------------------
Counter                                       | TM Name                   | Value
------------------------------------------------------------------------------------
capture.packets                               | Total                     | 11181144783
capture.rx_errors                             | Total                     | 48889667119
```

Reply:
It looks like `capture.packets` does not include the `dropped` packets, which is how it is in afpacket. So I guess `capture.packets` should include `imissed` and possible other errors?

Reply:
So, do you want to have one counter `capture.drop` that will include all packets that were dropped for whatever reason (whether it is RX or TX direction) and then individual counters like `capture.dpdk.rx_nombuf`, `capture.dpdk.imissed` and etc.?

Reply:
> It looks like `capture.packets` does not include the `dropped` packets, which is how it is in afpacket. So I guess `capture.packets` should include `imissed` and possible other errors?

I can reimplement the logic to follow the behavior of AFP.

Reply:
I have added the count of received packets to the total count of dropped packets. Here is what I currently have:
```c
if (ptv->queue_id == 0) {
        StatsSetUI64(ptv->tv, ptv->capture_dpdk_rx_errs,
                eth_stats.imissed + eth_stats.ierrors + eth_stats.rx_nombuf + ptv->pkts);
        StatsSetUI64(ptv->tv, ptv->capture_dpdk_tx_errs, eth_stats.oerrors);
        SC_ATOMIC_SET(ptv->livedev->drop,
                eth_stats.imissed + eth_stats.ierrors + eth_stats.rx_nombuf + eth_stats.oerrors + ptv->pkts);
    }
```

What I suggest is the following:
- remove the counter ptv->capture_dpdk_rx_errs and have independent counters for each kind of error
- the counter `ptv->livedev->drop` should only include packets that were received or failed to receive (it should not include `eth_stats.oerrors`) -> I haven't seen in AFP capture interface that it counts packets that were failed to a dropped (or any other) counter

Do you agree with the proposed solution?